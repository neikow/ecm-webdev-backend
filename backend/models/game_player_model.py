import enum

from nanoid import generate as generate_nanoid
from sqlmodel import SQLModel, Field


class UserRole(str, enum.Enum):
    admin = "admin"
    player = "player"


class GamePlayerModel(SQLModel, table=True):
    id: str = Field(default_factory=lambda: generate_nanoid(), primary_key=True)
    user_name: str
    room_id: int
    role: UserRole = Field(default=UserRole.player, index=True)
