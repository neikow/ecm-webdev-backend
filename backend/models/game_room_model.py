import enum
from datetime import datetime

from sqlalchemy import Column, Enum
from sqlmodel import SQLModel, Field


class GameType(str, enum.Enum):
    connect_four = "connect_four"


class GameRoomModel(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    created_at: datetime | None = Field(default_factory=lambda: datetime.now())
    password: str
    game_type: GameType = Field(
        sa_column=Column(Enum(GameType)), default=GameType.connect_four
    )
    is_active: bool = Field(default=True)
