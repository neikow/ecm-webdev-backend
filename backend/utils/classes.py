import inspect


def get_non_abstract_subclasses(cls: type) -> set[type]:
    """Recursively get all non-abstract subclasses of a class."""
    subclasses = set()
    for subclass in cls.__subclasses__():
        print(subclass.__name__, inspect.isabstract(subclass))
        if not inspect.isabstract(subclass):
            subclasses.add(subclass)

        subclasses.update(get_non_abstract_subclasses(subclass))

    return subclasses
