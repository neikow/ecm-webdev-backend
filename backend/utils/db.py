from typing import Generator

from sqlmodel import SQLModel, create_engine, Session

sqlite_file_name = "database.db"
sqlite_url = f"sqlite:///{sqlite_file_name}"
connect_args = {"check_same_thread": False}
engine = create_engine(sqlite_url, connect_args=connect_args)


def create_db_and_tables() -> None:
    SQLModel.metadata.create_all(engine)


def get_session() -> Generator[Session, None, None]:
    with Session(engine) as session:
        create_db_and_tables()
        yield session
