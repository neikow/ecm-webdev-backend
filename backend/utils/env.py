import os
from typing import Literal

from dotenv import load_dotenv


def get_env(key: Literal["CORS_ORIGINS", "JWT_SECRET_KEY", "BACKEND_COOKIE_DOMAIN", "DEV"]) -> str:
    load_dotenv()
    value = os.getenv(key)
    if value is None:
        raise ValueError(f"Environment variable {key} is not set.")
    return value
