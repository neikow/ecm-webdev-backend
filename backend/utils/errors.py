import enum


class ErrorCode(str, enum.Enum):
    ALREADY_IN_GAME_ROOM = "already_in_game_room"
    NOT_IN_GAME_ROOM = "not_in_game_room"
