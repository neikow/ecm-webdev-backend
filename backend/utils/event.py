from starlette.websockets import WebSocket

from backend.domain.events import BaseEvent
from backend.routers.websocket_schemas import WSMessageEvent


async def send_ws_message_event(ws: WebSocket, event: BaseEvent) -> None:
    await ws.send_json(
        WSMessageEvent(
            type="event",
            seq=event.seq,
            event=event
        ).model_dump()
    )
