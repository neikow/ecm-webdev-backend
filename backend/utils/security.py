from datetime import timezone, datetime, timedelta
from typing import Annotated

import jwt
from fastapi import Cookie
from fastapi.security import OAuth2PasswordBearer
from jwt.exceptions import DecodeError
from pydantic import BaseModel
from starlette.responses import Response

from backend.models.game_player_model import GamePlayerModel
from backend.utils.env import get_env

JWT_SECRET_KEY = get_env("JWT_SECRET_KEY")
ALGORITHM = "HS256"
AUTHORIZATION_COOKIE = "authorization"


class FailedToDecodeToken(Exception):
    pass


class TokenExpiredError(Exception):
    pass


oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")


class Token(BaseModel):
    access_token: str
    token_type: str


def create_access_token(data: GamePlayerModel, expires_delta: timedelta | None = None) -> str:
    to_encode = data.model_dump()
    if expires_delta:
        expire = datetime.now(timezone.utc) + expires_delta
    else:
        expire = datetime.now(timezone.utc) + timedelta(minutes=15)
    to_encode.update({"exp": expire})
    return jwt.encode(to_encode, JWT_SECRET_KEY, algorithm=ALGORITHM)


def verify_token(token: str) -> GamePlayerModel:
    try:
        payload = jwt.decode(token, JWT_SECRET_KEY, algorithms=[ALGORITHM])
        return GamePlayerModel(**payload)
    except jwt.ExpiredSignatureError:
        raise TokenExpiredError
    except DecodeError:
        raise FailedToDecodeToken


def add_authorization_cookie(response: Response, token: str) -> None:
    response.set_cookie(AUTHORIZATION_COOKIE, token, httponly=True)


def remove_authorization_cookie(response: Response) -> None:
    response.set_cookie(AUTHORIZATION_COOKIE, expires=0, max_age=0, httponly=True)


async def current_player_data(authorization: Annotated[str | None, Cookie()] = None) -> GamePlayerModel | None:
    if not authorization:
        return None
    try:
        return verify_token(authorization)
    except TokenExpiredError:
        return None
