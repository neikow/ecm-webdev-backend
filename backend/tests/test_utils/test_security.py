from unittest.mock import MagicMock

import pytest
import time_machine

from backend.models.game_player_model import GamePlayerModel, UserRole
from backend.utils.security import (
    create_access_token,
    verify_token,
    current_player_data, add_authorization_cookie, AUTHORIZATION_COOKIE, remove_authorization_cookie
)


@time_machine.travel(0.0)
def test_create_access_token(mocked_env, snapshot):
    with mocked_env({"JWT_SECRET_KEY": "test"}):
        token_data = GamePlayerModel(role=UserRole.admin, room_id=1, id="nanoid")
        encoded_token = create_access_token(token_data)
        assert encoded_token is not None
        assert encoded_token == snapshot


# Precomputed token for data: {"role": "admin", "room_id": 1, "id": "nanoid", "exp": 900}
ENCODED_TOKEN = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6Im5hbm9pZCIsInJvb21faWQiOjEsInJvbGUiOiJhZG1pbiIsImV4cCI6OTAwfQ.YLz0ED-y0w9BtLFTdT2tNX9XD8VkTwXwREDTwtWleTU"


@time_machine.travel(0.0)
def test_verify_token(mocked_env):
    with mocked_env({"JWT_SECRET_KEY": "test"}):
        verified_data = verify_token(ENCODED_TOKEN)
        assert verified_data.role == UserRole.admin
        assert verified_data.room_id == 1


@pytest.mark.asyncio
async def test_current_player_data_no_token():
    user = await current_player_data()
    assert user is None


@pytest.mark.asyncio
@time_machine.travel(0.0)
async def test_current_player_data_with_token():
    user = await current_player_data(authorization=ENCODED_TOKEN)
    assert user is not None
    assert user.role == UserRole.admin
    assert user.room_id == 1


def test_add_authorization_cookie() -> None:
    response = MagicMock()
    response.set_cookie = MagicMock()
    token = ENCODED_TOKEN
    add_authorization_cookie(response, token)
    response.set_cookie.assert_called_once_with(AUTHORIZATION_COOKIE, ENCODED_TOKEN, httponly=True)


def test_remove_authorization_cookie() -> None:
    response = MagicMock()
    response.set_cookie = MagicMock()
    remove_authorization_cookie(response)
    response.set_cookie.assert_called_once_with(AUTHORIZATION_COOKIE, expires=0, max_age=0, httponly=True)
