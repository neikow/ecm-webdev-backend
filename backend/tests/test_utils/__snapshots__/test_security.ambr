# serializer version: 1
# name: test_create_access_token
  'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6Im5hbm9pZCIsInJvb21faWQiOjEsInJvbGUiOiJhZG1pbiIsImV4cCI6OTAwfQ.YLz0ED-y0w9BtLFTdT2tNX9XD8VkTwXwREDTwtWleTU'
# ---
